OsmAnd-java/src/main/java/net/osmand/search/core/SearchCoreFactory.java

```diff
@@ -417,10 +418,19 @@ private void initAndSearchCities(final SearchPhrase phrase, final SearchResultMa
 		
 		private void searchPoiInCity(SearchPhrase nphrase, SearchResult res, SearchResultMatcher resultMatcher) throws IOException {
 			if (nphrase != null && res.objectType == ObjectType.CITY) {
+				int cntBefore = resultMatcher.getCount();
 				SearchAmenityByNameAPI poiApi = new SearchCoreFactory.SearchAmenityByNameAPI();
 				SearchPhrase newPhrase = nphrase.generateNewPhrase(nphrase, res.file);
-				newPhrase.getSettings().setOriginalLocation(res.location);
+				LatLon oldLocation = nphrase.getSettings().getOriginalLocation();
+				newPhrase.setOriginalLocation(res.location);//city location
+				newPhrase.addCityName(res.localeName);
+
 				poiApi.search(newPhrase, resultMatcher);
+
+				res.requiredSearchPhrase.setOriginalLocation(oldLocation);
+				if (resultMatcher.getCount() > cntBefore) {
+					res.requiredSearchPhrase.addKnownCity(res, nphrase.getUnknownWordToSearch());
+				}
 			}
```

Comment:
Search POI in SearchAddressByNameAPI when found city in searching.
In this case we change location search to city location ( bug in old code ```newPhrase.getSettings().setOriginalLocation(res.location);```)
Add city name for dynamicTagGroupMatcher
After searching come back old location search and store city for PoiUiFilter (Show on the map function).